### ISSUE TITLE
List Panic

### ISSUE THREAD

**Issue by stuartmscott** [fyne-io/fyne#1320]
### Describe the bug:
When a list displays fewer items than would fill the viewport, it releases unused items without a bounds check

```
panic: runtime error: index out of range [-1]

goroutine 15 [running]:
fyne.io/fyne/widget.(*listRenderer).Layout(0xc000170680, 0x318, 0x22f)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:127 +0x63b
fyne.io/fyne/widget.(*listRenderer).Refresh(0xc000170680)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:155 +0x54
fyne.io/fyne/widget.(*BaseWidget).Refresh(0xc00022c480)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/widget.go:137 +0x5c
```

```go
if l.visibleItemCount > 0 && l.list.Length() < l.visibleItemCount && len(l.children) > 0 { //<-- actual bounds check for children
	for i := l.visibleItemCount; i >= l.list.Length(); i-- {                           //<-- expected bounds check for children
		l.itemPool.Release(l.children[len(l.children)-1])
		l.children = l.children[:len(l.children)-1]                                //<-- children modification
	}
}
```

### To Reproduce:
Steps to reproduce the behaviour:
1. Create a dataset with one element
2. Create and display a list using the dataset
3. Add an element to the dataset
4. Call list.Refresh

### Device (please complete the following information):
 - OS: MacOS
 - Version: Catalina
 - Go version: go1.15 darwin/amd64
 - Fyne version: d12ce856aa1e477c8492dd503fc30856caf1a8fa

**Comment by andydotxyz** [comment]
Easier to replicate than that...

* Create a dataset with 0 elements
* Refresh the list

**Comment by Jacalz** [comment]
I think that I am running into the same issue on my end. I am using an empty list where items are added when they are created by the user:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6a1da9]

goroutine 1 [running, locked to thread]:
fyne.io/fyne/widget.(*listLayout).MinSize(0xc0001c8020, 0x0, 0x0, 0x0, 0x8ad940, 0x2)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:411 +0x29
fyne.io/fyne.NewContainerWithLayout(0x999d40, 0xc0001c8020, 0x0, 0x0, 0x0, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/container.go:138 +0x9b
fyne.io/fyne/widget.(*List).CreateRenderer(0xc0000d2580, 0x0, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:49 +0xbd
fyne.io/fyne/internal/cache.Renderer(0x9a3760, 0xc0000d2580, 0xc0000d2580, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/internal/cache/widget.go:29 +0x119
fyne.io/fyne/widget.(*BaseWidget).MinSize(0xc0000d2580, 0x9a2d60, 0xc0000d2580)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/widget.go:84 +0x45
fyne.io/fyne/widget.(*List).MinSize(0xc0000d2580, 0x0, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:60 +0x4c
fyne.io/fyne/layout.(*boxLayout).MinSize(0xc0000b1d7c, 0xc00009bcb0, 0x3, 0x3, 0x87d500, 0xc000099901)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/layout/boxlayout.go:152 +0x1df
fyne.io/fyne.NewContainerWithLayout(0x999b00, 0xc0000b1d7c, 0xc00009bcb0, 0x3, 0x3, 0xc0000d2630)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/container.go:138 +0x9b
main.(*appData).sendTab(0xc00009b680, 0x9a4a60)
        /home/jacob/go/src/github.com/Jacalz/wormhole-gui/send.go:88 +0x45a
main.main()
        /home/jacob/go/src/github.com/Jacalz/wormhole-gui/main.go:39 +0x1e5
```

**Comment by okratitan** [comment]
> I think that I am running into the same issue on my end. I am using an empty list where items are added when they are created by the user:
> 
> ```
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6a1da9]
> 
> goroutine 1 [running, locked to thread]:
> fyne.io/fyne/widget.(*listLayout).MinSize(0xc0001c8020, 0x0, 0x0, 0x0, 0x8ad940, 0x2)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:411 +0x29
> fyne.io/fyne.NewContainerWithLayout(0x999d40, 0xc0001c8020, 0x0, 0x0, 0x0, 0x0)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/container.go:138 +0x9b
> fyne.io/fyne/widget.(*List).CreateRenderer(0xc0000d2580, 0x0, 0x0)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:49 +0xbd
> fyne.io/fyne/internal/cache.Renderer(0x9a3760, 0xc0000d2580, 0xc0000d2580, 0x0)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/internal/cache/widget.go:29 +0x119
> fyne.io/fyne/widget.(*BaseWidget).MinSize(0xc0000d2580, 0x9a2d60, 0xc0000d2580)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/widget.go:84 +0x45
> fyne.io/fyne/widget.(*List).MinSize(0xc0000d2580, 0x0, 0x0)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:60 +0x4c
> fyne.io/fyne/layout.(*boxLayout).MinSize(0xc0000b1d7c, 0xc00009bcb0, 0x3, 0x3, 0x87d500, 0xc000099901)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/layout/boxlayout.go:152 +0x1df
> fyne.io/fyne.NewContainerWithLayout(0x999b00, 0xc0000b1d7c, 0xc00009bcb0, 0x3, 0x3, 0xc0000d2630)
>         /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/container.go:138 +0x9b
> main.(*appData).sendTab(0xc00009b680, 0x9a4a60)
>         /home/jacob/go/src/github.com/Jacalz/wormhole-gui/send.go:88 +0x45a
> main.main()
>         /home/jacob/go/src/github.com/Jacalz/wormhole-gui/main.go:39 +0x1e5
> ```
> 
> https://gist.github.com/Jacalz/c1aa28be0536f602f631845db669704d

As we talked about on Slack - This was a result of extending the list widget without providing the proper methods for Length(), CreateItem(), and UpdateItem() it would appear.

**Comment by okratitan** [comment]
Fixed By https://github.com/fyne-io/fyne/pull/1323

**Comment by stuartmscott** [comment]
```go

type AList struct {
	widget.List
	ids        []string
}

func NewAList(callback func(id string)) *AList {
	l := &AList{
		List: widget.List{
			CreateItem: func() fyne.CanvasObject {
				return &widget.Label{}
			},
		},
	}
	l.Length = func() int {
		return len(l.ids)
	}
	l.UpdateItem = func(index int, item fyne.CanvasObject) {
		name := l.ids[index]
		if name == "" {
			name = "(untitled)"
		}
		item.(*widget.Label).SetText(name)
	}
	l.OnItemSelected = func(index int) {
		log.Println("Selected", index)
	}
	l.ExtendBaseWidget(l)
	return l
}

func (l *AList) Append(id string) {
	l.ids = append(l.ids, id)
}
```
```
09-18 18:48:20.692  3648     0 E Go      : panic: runtime error: index out of range [-1]
09-18 18:48:20.692  3648     0 E Go      : 
09-18 18:48:20.692  3648     0 E Go      : goroutine 51 [running]:
09-18 18:48:20.692  3648  3679 E GoLog   : panic: runtime error: index out of range [-1]
09-18 18:48:20.692  3648     0 E Go      : fyne.io/fyne/widget.(*listRenderer).Layout(0x4000308300, 0x160, 0x218)
09-18 18:48:20.692  3648  3679 E GoLog   : 
09-18 18:48:20.692  3648     0 E Go      : 	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:127 +0x5a0
09-18 18:48:20.692  3648     0 E Go      : fyne.io/fyne/internal/driver/gomobile.updateLayout(0x79aea31400, 0x400032c000)
09-18 18:48:20.692  3648     0 E Go      : 	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/internal/driver/gomobile/canvas.go:275 +0x108
```

```
$ fyne version
fyne cli version v1.3.4-0.20200918231643-0a2fa00d4f28
```

**Comment by stuartmscott** [comment]
While trying to find a simple reproduction case, I found removing from the slice can cause issues - add this towards the bottom of fyne_demo list
```
$ git diff
diff --git a/cmd/fyne_demo/screens/widget.go b/cmd/fyne_demo/screens/widget.go
index fc777b7e..7691e563 100644
--- a/cmd/fyne_demo/screens/widget.go
+++ b/cmd/fyne_demo/screens/widget.go
@@ -314,6 +314,15 @@ func makeListTab() fyne.CanvasObject {
                label.SetText(data[index])
                icon.SetResource(theme.DocumentIcon())
        }
+
+       go func() {
+               for len(data) > 0 {
+                       data = data[1:]
+                       list.Refresh()
+                       time.Sleep(time.Second)
+               }
+       }()
+
        split := widget.NewHSplitContainer(list, fyne.NewContainerWithLayout(layout.NewCenterLayout(), hbox))
        return fyne.NewContainerWithLayout(layout.NewMaxLayout(), split)
 }
```
Then scroll to the bottom of the list and get
```$ go fmt ./... && go run cmd/fyne_demo/main.go
panic: runtime error: index out of range [985] with length 985

goroutine 34 [running]:
fyne.io/fyne/cmd/fyne_demo/screens.makeListTab.func3(0x3d9, 0x445b900, 0xc000180640)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/cmd/fyne_demo/screens/widget.go:310 +0xcc
fyne.io/fyne/widget.(*listRenderer).Layout(0xc00057e200, 0x142, 0x1ac)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:142 +0x4f8
fyne.io/fyne/widget.(*listRenderer).Refresh(0xc00057e200)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:156 +0x54
fyne.io/fyne/widget.(*BaseWidget).Refresh(0xc00021f0e0)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/widget.go:137 +0x5c
fyne.io/fyne/cmd/fyne_demo/screens.makeListTab.func5(0xc00000e920, 0xc00021f0e0)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/cmd/fyne_demo/screens/widget.go:321 +0x33
created by fyne.io/fyne/cmd/fyne_demo/screens.makeListTab
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/cmd/fyne_demo/screens/widget.go:318 +0x3ff
exit status 2
```

**Comment by stuartmscott** [comment]
My bad. The package name had changed so I was launching the old app that didn't have fix #1323
